app/backend/src/couchers/models.py

```diff
@@ -655,23 +625,113 @@ class HostRequest(Base):
     from_user_id = Column(ForeignKey("users.id"), nullable=False, index=True)
     to_user_id = Column(ForeignKey("users.id"), nullable=False, index=True)
 
-    # dates as "YYYY-MM-DD", in the timezone of the host
-    from_date = Column(String, nullable=False)
-    to_date = Column(String, nullable=False)
+    # TODO: proper timezone handling
+    timezone = "Etc/UTC"
+
+    # dates in the timezone above
+    from_date = Column(Date, nullable=False)
+    to_date = Column(Date, nullable=False)
 
     status = Column(Enum(HostRequestStatus), nullable=False)
 
     to_last_seen_message_id = Column(BigInteger, nullable=False, default=0)
     from_last_seen_message_id = Column(BigInteger, nullable=False, default=0)
 
+    start_time_to_write_reference = column_property(date_in_timezone(to_date, timezone))
+    end_time_to_write_reference = column_property(date_in_timezone(to_date, timezone) + text("interval '14 days'"))
+
     from_user = relationship("User", backref="host_requests_sent", foreign_keys="HostRequest.from_user_id")
     to_user = relationship("User", backref="host_requests_received", foreign_keys="HostRequest.to_user_id")
     conversation = relationship("Conversation")
 
+    @hybrid_property
+    def can_write_reference(self):
+        return (
+            (self.status == HostRequestStatus.confirmed)
+            & (now() >= self.start_time_to_write_reference)
+            & (now() <= self.end_time_to_write_reference)
+        )
+
+    @can_write_reference.expression
+    def can_write_reference(cls):
+        return (
+            (cls.status == HostRequestStatus.confirmed)
+            & (func.now() >= cls.start_time_to_write_reference)
+            & (func.now() <= cls.end_time_to_write_reference)
+        )
+
     def __repr__(self):
         return f"HostRequest(id={self.id}, from_user_id={self.from_user_id}, to_user_id={self.to_user_id}...)"
 
 
+class ReferenceType(enum.Enum):
+    friend = enum.auto()
+    surfed = enum.auto()  # The "from" user surfed with the "to" user
+    hosted = enum.auto()  # The "from" user hosted the "to" user
+
+
+class Reference(Base):
+    """
+    Reference from one user to another
+    """
+
+    __tablename__ = "references"
+
+    id = Column(BigInteger, primary_key=True)
+    # timezone should always be UTC
+    time = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
+
+    from_user_id = Column(ForeignKey("users.id"), nullable=False, index=True)
+    to_user_id = Column(ForeignKey("users.id"), nullable=False, index=True)
+
+    reference_type = Column(Enum(ReferenceType), nullable=False)
+
+    host_request_id = Column(ForeignKey("host_requests.id"), nullable=True)
+
+    text = Column(String, nullable=True)  # plain text
+
+    rating = Column(Float, nullable=False)
+    was_appropriate = Column(Boolean, nullable=False)
+
+    visible_from = Column(DateTime(timezone=True), nullable=False)
```

Comment:
Can be calculated from `time`, no need to store in db.

Reply:
The reason I've added it is that it's really quite annoying to compute. Logic is approximately this:

1. If it's a friend reference, it's immediately visible
2. Otherwise, it's visible if the other reference from the person you surfed/hosted has written the reference OR if the reference is over 2 weeks old

Storing this means it's very easy to get all visible references, avoiding a few hairy joins in some places. But I can also remove this if you feel it's important.

Reply:
I'll change it.

But this is set once, except for a hosting/surifng reference if the other person writes a reference within 2 weeks. Doesn't get updated if you friend/unfriend/tz change,etc.

Reply:
Well if visible_from should be based on an old snapshot of the database then we can't calculate it on demand, then we need a column for it. Just have to make up our mind :D

Reply:
Done, turned into a subquery: I agree, it's cleaner this way! Thanks :)